### ISSUE TITLE
Facing issue in running "conf/dpdk/dpdk-bounce.click" example configuration

### ISSUE THREAD

**Issue by maulikpatel1990** [tbarbette/fastclick#178]
I compiled fastclick with dpdk support in Ubuntu 16.04 (4.15.0-47-generic) without any issues but when I try to launch sample config using below command, I am not able to run the same. Below is the command along with its output:

_maulik@maulik-VPCEB14EN:~/projects/click/fastclick$ sudo bin/click --dpdk  -- conf/dpdk/dpdk-bounce.click
EAL: Detected 4 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Probing VFIO support...
conf/dpdk/dpdk-bounce.click:11: While configuring 'FromDPDKDevice@1 :: FromDPDKDevice':
  0: Unknown or invalid PORT
conf/dpdk/dpdk-bounce.click:14: While configuring 'ToDPDKDevice@4 :: ToDPDKDevice':
  0 : Unknown or invalid PORT
Router could not be initialized!_

I tried to ran some other examples also but I am getting similar error like "  0 : Unknown or invalid PORT", any hints what I am missing at my end?

Thanks in advance

Regards,
Maulik

**Comment by tbarbette** [comment]
Hi,

What NIC are you using ? Intel NICs must be bound to the igb_uio driver as per DPDK documentation. Mellanox NICs must be explicitely enabled when building DPDK.

Tom

**Comment by maulikpatel1990** [comment]
Hi Tom,

Thanks for your response. I am trying to run everything on Ubuntu 16.04 user space environment. Is there is any specific requirement of Intel/Mellanox NIC to run DPDK?
 
Below are my NIC details:
```
maulik@maulik-VPCEB14EN:~$ sudo lshw -c network
[sudo] password for maulik: 
  *-network               
       description: Wireless interface
       product: AR9285 Wireless Network Adapter (PCI-Express)
       vendor: Qualcomm Atheros
       physical id: 0
       bus info: pci@0000:02:00.0
       logical name: wlp2s0
       version: 01
       serial: 78:dd:08:d4:a5:39
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list ethernet physical wireless
       configuration: broadcast=yes driver=ath9k driverversion=4.15.0-47-generic firmware=N/A ip=192.168.0.103 latency=0 link=yes multicast=yes wireless=IEEE 802.11
       resources: irq:16 memory:f4a00000-f4a0ffff
  *-network
       description: Ethernet interface
       product: Yukon Optima 88E8059 [PCIe Gigabit Ethernet Controller with AVB]
       vendor: Marvell Technology Group Ltd.
       physical id: 0
       bus info: pci@0000:04:00.0
       logical name: enp4s0
       version: 11
       serial: 54:42:49:0e:37:7d
       capacity: 1Gbit/s
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list rom ethernet physical tp 10bt 10bt-fd 100bt 100bt-fd 1000bt 1000bt-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=sky2 driverversion=1.30 latency=0 link=no multicast=yes port=twisted pair
       resources: irq:24 memory:f2220000-f2223fff ioport:a000(size=256) memory:f2200000-f221ffff
```

I am running following set of commands on my terminal as per my knowledge:
```
maulik@maulik-VPCEB14EN:~$ sudo sh -c 'echo 1024 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages'
[sudo] password for maulik: 
maulik@maulik-VPCEB14EN:~$ sudo mkdir -p /mnt/huge
maulik@maulik-VPCEB14EN:~$ sudo mount -t hugetlbfs nodev /mnt/huge
maulik@maulik-VPCEB14EN:~$ export DPDK_DIR=/home/maulik/projects/click/dpdk/
maulik@maulik-VPCEB14EN:~$ cd $DPDK_DIR
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo modprobe uio
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo insmod x86_64-native-linuxapp-gcc/kmod/igb_uio.ko
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ $DPDK_DIR/usertools/dpdk-devbind.py --status

Network devices using kernel driver
===================================
0000:02:00.0 'AR9285 Wireless Network Adapter (PCI-Express) 002b' if=wlp2s0 drv=ath9k unused=igb_uio *Active*
0000:04:00.0 'Yukon Optima 88E8059 [PCIe Gigabit Ethernet Controller with AVB] 4381' if=enp4s0 drv=sky2 unused=igb_uio 

No 'Crypto' devices detected
============================

No 'Eventdev' devices detected
==============================

No 'Mempool' devices detected
=============================

No 'Compress' devices detected
==============================
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo $DPDK_DIR/usertools/dpdk-devbind.py --bind=igb_uio 04:00.0
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo $DPDK_DIR/usertools/dpdk-devbind.py --bind=igb_uio 02:00.0
Routing table indicates that interface 0000:02:00.0 is active. Not modifying
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo $DPDK_DIR/usertools/dpdk-devbind.py --bind=igb_uio 02:00.0
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ $DPDK_DIR/usertools/dpdk-devbind.py --status

Network devices using DPDK-compatible driver
============================================
0000:02:00.0 'AR9285 Wireless Network Adapter (PCI-Express) 002b' drv=igb_uio unused=ath9k
0000:04:00.0 'Yukon Optima 88E8059 [PCIe Gigabit Ethernet Controller with AVB] 4381' drv=igb_uio unused=sky2

No 'Crypto' devices detected
============================

No 'Eventdev' devices detected
==============================

No 'Mempool' devices detected
=============================

No 'Compress' devices detected
==============================
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ sudo $DPDK_DIR/examples/helloworld/build/helloworld -c f -n 4
EAL: Detected 4 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Probing VFIO support...
hello from core 1
hello from core 2
hello from core 3
hello from core 0
maulik@maulik-VPCEB14EN:~/projects/click/dpdk$ cd /home/maulik/projects/click/fastclick
maulik@maulik-VPCEB14EN:~/projects/click/fastclick$ sudo bin/click --dpdk -- conf/dpdk/dpdk-bounce.click
EAL: Detected 4 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Probing VFIO support...
conf/dpdk/dpdk-bounce.click:11: While configuring 'FromDPDKDevice@1 :: FromDPDKDevice':
  0: Unknown or invalid PORT
conf/dpdk/dpdk-bounce.click:14: While configuring 'ToDPDKDevice@4 :: ToDPDKDevice':
  0 : Unknown or invalid PORT
Router could not be initialized!
maulik@maulik-VPCEB14EN:~/projects/click/fastclick$
```
Am I missing something here? How can I find port assigned by DPDK? Let me know in case you need any additional information?

Regards,
Maulik

**Comment by tbarbette** [comment]
I don't think yours are compatible, unfortunately :
https://core.dpdk.org/supported/

But does your project require high speed? If it's just about testing/trying something, then good old FromDevice is fine.

Tom

**Comment by maulikpatel1990** [comment]
Tom, Thanks again for looking into this.

As of now I just want to run fastclick with DPDK and want to explore some stuffs on DPDK. Any pointers from where can I start looking into?
Any sample configs for the same?

**Comment by tbarbette** [comment]
https://github.com/tbarbette/fastclick/blob/master/conf/dpdk/dpdk-bounce.click would be the simplest example there is.

The rest is Click specific. A start would be https://github.com/tbarbette/fastclick/wiki/Tutorial

For your NIC problem, you may use a "fake" driver such as AF_PACKET. It will have correct performance and will fully behave as a DPDK driver giving DPDK packets : https://doc.dpdk.org/guides/nics/af_packet.html 

Basically just run fastclick with click --dpdk --vdev=eth_af_packet0,iface=eth0,blocksz=4096,framesz=2048,framecnt=512,qpairs=1,qdisc_bypass=0 -- config.click
given that eth0 is your NIC ifname.

**Comment by gkatsikas** [comment]
You could also try DPDK KNI if this is possible with your NIC. This way you eliminate some of the bottlenecks of regular Linux interfaces, while retaining useful Linux tools such as ethtool
